### PR TITLE
[WIP] Implemented infix mathematical operators through abstract type `:number`

### LINKED ISSUE #1757: Support mathematical infix operators
I was wondering if there was any interest with adding common mathematical infix operators (+, -, *, and /) to the Query API.

I am interested in knowing if this is desired and would be accepted as a contribution.

### PR DESCRIPTION
**Pull request by costaraphael**
Fixes #1757 

I gave it a shot and came up with this.

Somethings I'd like to question:

- I needed to add a explicit type cast in the PG integration test (the `#FIXME` tag). Is it an expected behavior?
- I'm not very fond of how the number type checking was implemented at the builder stage. Do you have any tips on how it could be better implemented?
- I have a feeling that there are some missing test cases. Are there some edge cases you think I should test?

I decided to validate my approach first before implementing the MySQL driver.

I'm looking forward for your feedback! :smiley:

### CODE REVIEW

File: lib/ecto/query/builder.ex [hunk 1]
```diff
@@ -178,6 +180,26 @@ defmodule Ecto.Query.Builder do
      {params |> wrap_nil(left) |> wrap_nil(right), acc}}
   end
 
+  # mathematical operators
+  def escape({math_op, _, [left, right]} = expr, type, params_acc, vars, env)
+      when math_op in ~w(+ - * /)a do
+    assert_type!(expr, type, :number)
```
Comment (josevalim) [inline lib/ecto/query/builder.ex]: I don't think you need this call, do you? Since we are checking the type below.

Comment (josevalim) [inline lib/ecto/query/builder.ex]: Actually, ignore me, this is correct.

File: lib/ecto/query/builder.ex [hunk 2]
```diff
@@ -178,6 +180,26 @@ defmodule Ecto.Query.Builder do
...
+    
+    args_type =
```
Comment (josevalim) [inline lib/ecto/query/builder.ex]: I think `args_type` should always be `:number` and then we handle the case of `:number` later on quoted_type implementation.

File: lib/ecto/query/builder.ex [hunk 3]
```diff
@@ -661,6 +687,18 @@ defmodule Ecto.Query.Builder do
   def quoted_type(literal, _vars) when is_boolean(literal), do: :boolean
   def quoted_type(literal, _vars) when is_integer(literal), do: :integer
 
+  # Mathematical operations
+  def quoted_type({:/, _, [_, _]}, _vars), do: :float
+  def quoted_type({math_op, _, [_, _] = args}, vars) when math_op in ~w(+ - *)a do
+    case Enum.map(args, &quoted_type(&1, vars)) do
+      [:integer, :integer]           -> :integer
+      [:decimal, :decimal]           -> :decimal
+      [a, b] when a in @number_types
```
Comment (josevalim) [inline lib/ecto/query/builder.ex]: Does `integer + decimal` return `float`?

File: integration_test/cases/type.exs [hunk 4]
```diff
@@ -26,11 +26,16 @@ defmodule Ecto.Integration.TypeTest do
...
     assert [1] = TestRepo.all(from p in Post, where: p.visits == 1, select: p.visits)
+    assert [3] = TestRepo.all(from p in Post, select: p.visits + 2)
 
     # Floats
     assert [0.1] = TestRepo.all(from p in Post, where: p.intensity == ^float, select: p.intensity)
     assert [0.1] = TestRepo.all(from p in Post, where: p.intensity == 0.1, select: p.intensity)
     assert [1500.0] = TestRepo.all(from p in Post, select: 1500.0)
+    assert [1.1] = TestRepo.all(from p in Post, select: p.intensity + ^integer)
+    assert [1.1] = TestRepo.all(from p in Post, select: p.visits + 0.1)
+    # FIXME There should be no need for typecasting (?)
```
Comment (josevalim) [inline integration_test/cases/type.exs]: If this is not allowed without type casting, the ideal would be to not allow mixing types. Always integer | decimal + integer | decimal or float + float.

**Comment by josevalim** [comment]
Hi @costaraphael! I have added some comments but I am not confident about any of them. If you are having typing issues, maybe the best solution is to limit the interoperability of the arithmetic operators?

Maybe we should try the different combinations without the query syntax first. For example:

    TestRepo.query("SELECT $1 + $2", [1, 1.0])
    TestRepo.query("SELECT $1 + $2", [1, Decimal.new("1.0")])
    TestRepo.query("SELECT $1 + $2", [1.0, Decimal.new("1.0")])
    ...

Which of those will pass? This can help us figure out our own typing rules.

**Comment by costaraphael** [comment]
Hi @josevalim!

So, about interoperability, I was thinking that when the operation has a float, it should always return a float, because of the lack of precision, this would leave us with the following cases:

```elixir
[:float, _] -> :float
[_. :float] -> :float
# there are no floats from now on
[:decimal, _] -> :decimal
[_, :decimal] -> :decimal
[type, type] -> type
```

But where should this logic go? In the builder it works really nice until you interpolate parameters in the query. One possible solution is that when using parameters in these operations you should always type cast them. What do you think?

**Comment by costaraphael** [comment]
@josevalim I added those tests and it looks like the database behavior is a lot linked with how the types are cast in the query.

**Comment by josevalim** [comment]
@costaraphael the issue is that Ecto types are a convenience but they are not precisely sound. In many places the type will be `:any` and you won't have enough information to check the left and right side of `+`. So the most important is to guarantee we have a syntax that works regardless of the operands. From your snippet, we can see that "SELECT $1 + $2" fails. But what about "SELECT $1::numeric + $2::numeric"? Could that work for all inputs?

If that works, that's probably the best way to go. To introduce a `:numeric` type in Ecto and make sure the operands and result of `+` are all numeric.

**Comment by costaraphael** [comment]
It works, but then the results will always be Decimal structs, even if the operation is a simple `1 + 3`.

If this behavior is acceptable then that's the way to go, way simpler than anything I could come up with.

**Comment by josevalim** [comment]
@costaraphael what happens if you use `$1::numeric + $2::numeric` and you pass a float? Also, what if you use `::numeric` but one of the operands is a field? will it complain?

**Comment by costaraphael** [comment]
@josevalim I'll need to run some more tests, but I'm pretty sure it works just fine, the only caveat being that when an operation like this has a float type field as an operand, it will not return a Decimal, it will return a float, regardless of the other operand

**Comment by josevalim** [comment]
So if I understand correctly, if any of the sides is a float, it returns a float, otherwise a decimal. Correct? Is this consistent across adapters?

**Comment by josevalim** [comment]
Btw, thanks for all of the investigative work. Often it is the most time consuming part of adding features to Ecto. :)

**Comment by costaraphael** [comment]
> So if I understand correctly, if any of the sides is a float, it returns a float, otherwise a decimal. Correct?

In a more correct statement, if any of the args is a float, the result is a float, then if any of the args is a numeric, the result is a decimal, and if both the args are integers, the result is an integer. Casting all params to `::numeric` and enveloping the expression in something like `(a + b)::numeric` the result of any combination of types should be decimal.

> Is this consistent across adapters?

I cannot say for the MySQL adapter, I'll run some more tests latter, but I guess it will be a lot less strict then PostgreSQL regarding of types. It will probably be more clumsy to enforce a return type, I guess.

> Btw, thanks for all of the investigative work. Often it is the most time consuming part of adding features to Ecto. :)

No worries man, I'm loving this. And I'll make sure this is not my last contribution to Ecto or to the Elixir ecosystem :)

**Comment by costaraphael** [comment]
@josevalim so, I spent some time testing on both databases, and the behavior is consistent: if any of the args is a float, the result is a float, then if any of the args is a numeric (or decimal in mysql), the result is a decimal, and if both the args are integers the result is an integer (this last one does not apply to division on mysql, where it does not perform integer division and it returns a decimal).

I also confirmed my expectations on the lack of strictness of mysql types.

In conclusion. casting both sides of the operation to numeric/decimal should work fine (even if one or both sides are not args, but fields or literals), and it would guarantee that the return type is always a decimal while making the query type checking a lot easier at the same time.

Should I go this way?

File: integration_test/cases/type.exs (hunk 4 above)
Comment (ericmj) [inline integration_test/cases/type.exs]: This is due to how determines the types for parameters. Postgresql sees `integer + ?` so it infers that `?` will also be an integer. If you pass a float the encoder will of course break when it expects an integer.

I don't think this should affect our decisions on mixing types for numeric operators since we have this issue for all queries run against postgres.

**Comment by josevalim** [comment]
Thanks @costaraphael! So to confirm:

  * In Postgres we will emit `(left)::numeric + (right)::numeric`
  * In MySQL we weil emit `left + right`
  * In Ecto we will introduce a pseudo type called `:numeric` and assert that + receives numeric and returns numeric

Implementation-wise, we will add a new `:numeric` type to Ecto.Type that if it receives an integer, it returns an integer, if it receives a float, it returns a float, if anything else, it acts like a decimal.

Do you think this is the way to go?

**Comment by josevalim** [comment]
Actually, please discard the last paragraph in my last comment. I think we should go with your current logic today where you implement the `[:float, _], [_, :float], [:decimal, _], [_, :decimal], [:integer, :integer], etc` lookup rules. This is going to be less complex than introducing a new `:numeric` type. You just need to be careful because many times we simply don't know the type. For example:

    from "posts", select: p.visits + 1

We don't know the type of visits, so we need to handle this case accordingly.

**Comment by costaraphael** [comment]
> In Postgres we will emit (left)::numeric + (right)::numeric

Correct

> In MySQL we weil emit left + right

Actually to have the same level of consistency as in Postgres, we should emit `CAST(left as decimal) + CAST(right as decimal)`. This way both databases should always return the same type.

> if it receives an integer, it returns an integer, if it receives a float, it returns a float, if anything else, it acts like a decimal

So, using the database casts in Postgres and MySQL, we will assure that all math infix operations return a Decimal, regardless of it's operands types. I know this may sound like a limitation, but the type lookup logic I mentioned earlier tends to get too complex. In some edge cases (using parameters or fields) it may be hard to predict what it should return. Say, in a situation like this:

```elixir
from p in "posts", select: p.visits + ^some_number
```

What type will the operation return? I think always returning a Decimal is way more predictable, simpler and gives a less confusing experience.

> Actually, please discard the last paragraph in my last comment. I think we should go with your current logic today where you implement the [:float, _], [_, :float], [:decimal, _], [_, :decimal], [:integer, :integer], etc lookup rules. This is going to be less complex than introducing a new :numeric type. You just need to be careful because many times we simply don't know the type.

I really don't know if this is the simpler way anymore, exactly because of the times where we don't know the type. In the example above, what type should I cast the argument to? That's the main complexity I faced when trying to implement this for the first time.

I'm thinking that the simplest approach is to try to cast any args in these operations to the `:decimal` Ecto type, and use the database level casts to ensure the result is always a Decimal struct.

Thoughts?

**Comment by josevalim** [comment]
@costaraphael I think returning a decimal though when you expect an integer will be confusing for users. On the other hand, not asserting the type is not a huge problem as long as we can properly cast it at the database level. In other words, it is OK to focus on expressiveness if it makes Ecto types less expressive.

Another option though is to not use `+` but to use integer_add, float_add and decimal_add and friends. This will allow us to always be type safe and the database will always get the proper types. This will also be helpful if a database does not support the same casting rules as PG and MySQL.

The downside is that it is slightly more verbose but it may be a reasonable trade-off. We can still check for `+` in the AST and raise reasonable error messages.

**Comment by costaraphael** [comment]
Hi @josevalim!

So, I put a lot of thought into this today, and I think I came up with an idea that may be more simple and terse to the end user while having a predictable behavior and return types through the adapters.

Up until now, we were focusing on defining what the operation should return based on the types of the operands. This was bound to be error prone and complex, as there are a lot of cases where we don't know the operands types.

What if we looked at the problem the other way around? We could make the user specify what he wants, and then cast the operands accordingly. This way, unidentified operand types would be no problem, they would just be casted at the `planner` phase. And the return types would be predictable, because the user will have to specify what he wants (integer, float or decimal).

The idea of implementing the `integer_add`, `float_add`, etc functions will achieve this goal, but although a simple add operation will not be that verbose (e.g. `float_add(1, 2)`), adding more operations can make it really verbose really fast (e.g. `float_div(float_add(1, 2), float_mul(2, 5))`).

So I came up with this API and split it into two parts, one for the query level (mostly compile time) and another one for the adapter level(ish) (warning, fake unimplemented API ahead):

#### Query level

We will need three functions in the query API: `integer/1`, `float/1` and `decimal/1`. These functions should not make to the database level. They should be striped away at compile time. The only thing that this functions do is enforcing that the expressions passed to them have these respective types: integer, float and decimal.

With these functions we could write queries like this:
```elixir
from p in Post, select: integer(p.visists + ^foo)
```
And whatever are the types of the operands, they should be cast to integer to perform an integer addition. So how is that better than a simple `integer_add/1` function? Simple, now that we know that the return type of the whole expression is an integer, we also know that both operands should be integers (or castable to integer), and we could chain more expressions like this:
```elixir
from p in Post, select: integer(p.visists + (^foo / ^bar))
```
We don't need to explicitly call `integer` again because we already know that the right side must be an integer, and as it's another expression, the logic just repeats recursively. This last example maps 1:1 to the following code, using the more verbose functions:
```elixir
from p in Post, select: integer_add(p.visists, integer_div(^foo, ^bar))
```

If some user attempts to use an infix operation where there's no information of what type the result should be (something like: `from p in Post, select: p.visits + 1`), we should raise an error asking the user to specify one of the three functions discussed above.

But this does not make it any easier on the adapter level, so that's where the other part kicks in.

#### Adapter level

To give us the same flexibility and predictability on the adapter level, the parsing itself of the of the infix operators should output some data structure containing the operator, what's the type of the operation and the two operands. Something like: `{:math_operation, :/, :integer, operand1, operand2}`.

Then, when building the SQL (or whatever the adapter uses to query stuff) the adapter could go like: "So, this is a division operation that must return an integer. How should I do this?".

This should also be compatible to the `integer_add` and friends functions approach. It's just that I think it will be easier to deal with a data structure like that, as both Postgres and MySQL could use the operator atom to dynamically build the SQL.

--

What do you think about this approach? Are there any constraints that I'm missing?

**Comment by josevalim** [comment]
Thanks @costaraphael.

My concern with the approach above is that it introduces another mechanism to cast the results. For example, should we expect `integer(p.some_float_column)` to work... or is this exclusive to arithmetic operations? Why don't we then use `type(a + b, :integer)`? Probably because it is more verbose.

On the other hand, I agree it solves the issues with `integer_add` and friends. I am unsure which approach would be best.

**Comment by costaraphael** [comment]
@josevalim I actually prefer to use `type/2` instead of the `integer/1`, `float/1` and `decimal/1` functions, as it keeps the query API untouched and makes the implementation way simpler. It may be slightly more verbose, but since the user will only need to use it when there's no way to know what the return type should be, I don't think it will be a problem.

But it seems that the `type/2` function [only works with query parameters](https://github.com/elixir-ecto/ecto/blob/fc288d14a27908c7f93d21776b64312b6b3e6316/lib/ecto/query/builder.ex#L51). We will need to change it to work with arithmetic operations too.

As the `type/2` function currently works only with query parameters, I guess we should just add support for arithmetic expressions and keep it that way for now. It would be awesome to use it to cast database values (`type(p.string_column, :integer)`), but this is way out of the scope of this PR.

**Comment by josevalim** [comment]
@costaraphael and it has the downside of being the most verbose proposal of them all :S

**Comment by costaraphael** [comment]
@josevalim what about we add the less verbose functions (`integer/1`, ...) as a simple syntactic sugar for the `type/2` function?

**Comment by josevalim** [comment]
@costaraphael that has the downsides mentioned above. So far we don't have an optimal solution... so I will probably have to sit with @michalmuskala/@ericmj and figure out what is the best option to move forward.

**Comment by OvermindDL1** [comment]
Would be nice to be able to do, say, add a date and an interval instead of needing to use fragments as I am having to do currently.  There are other types where `+` and so forth are valid as well.

**Comment by josevalim** [comment]
@OvermindDL1 we already have datetime_add I think?

**Comment by OvermindDL1** [comment]
> @OvermindDL1 we already have datetime_add I think?

I'm talking about the actual `postgresql` `interval` type.

**Comment by costaraphael** [comment]
@OvermindDL1 you mean something like: `p.some_date + p.some_interval`?

**Comment by OvermindDL1** [comment]
> @OvermindDL1 you mean something like: p.some_date + p.some_interval?

Precisely.  I have a table that has a lot of foreign tables linked to it, the main table has an interval and the subtables are checked for any that do not have a corresponding  sibling subtable within the specified interval on the main table, all done in SQL.  I use a lot of fragments...  ^.^;

**Comment by josevalim** [comment]
@OvermindDL1 that's very specific to Postgres so it is unlikely we will see it in Ecto. Also remember that you can wrap fragments in macros and that should make it feel as if it was part of Ecto (except by an import line).

**Comment by costaraphael** [comment]
@josevalim So, another approach is to pass the expressions to the database directly and let it handle them. The main problem with this is that the behavior would be less predictable just by looking at the code, as you would need to know the database schema to know what type an infix operation might return (something like `p.foo + p.bar`), and the behavior may differ from adapter to adapter (`1/2` returns `0` in Postgres and `#Decimal<0.5>` in MySQL). Also, we would probably need the user to explicitly use `type/2` on some query parameters while using these operations.

I know this is also not an ideal solution, but it may be worth considering it's advantages and implications, even if just to spark some new ideas. ^^

I'll be waiting for you guys to reach a consensus on what's the best way to go.

**Comment by hauleth** [comment]
I would go with last proposal. The whole point of Ecto in my opinion is to give thin wrapper on most common features but this doesn't free you from understanding your DB system.

**Comment by costaraphael** [comment]
> I would go with last proposal. The whole point of Ecto in my opinion is to give thin wrapper on most common features but this doesn't free you from understanding your DB system.
>
> EDIT:
>
> @costaraphael also if we will support this then when someone want to force result type then one can use type(a / b, :decimal) or type(a / b, :float).

I like that, and as time passes by, I grow more and more fond of this approach. I'm no core committer to be sure about this, but it feels like the "Ecto way" of doing things.

@josevalim What do you think? I have some work to do this weekend, but I could work on it this next week if this is the way to go.

**Comment by josevalim** [comment]
I agree with @costaraphael and @hauleth :)

**Comment by costaraphael** [comment]
@josevalim Awesome! I'll get working on this then :)

Should I open a new PR and close this one?

**Comment by josevalim** [comment]
I would start a new one because we can revisit this one if it turns out the new approach is not the best one. :)

**Comment by costaraphael** [comment]
Closing this one due to the fact that #2400 is already merged.
